Fix issues reported in ZD13606

julek-wolfssl opened the pull request:
- Ticket failure should result in a regular handshake
- Fix `BioReceive` for closed connection
  -The pending check was forcing a `WOLFSSL_CBIO_ERR_WANT_WRITE` return even though the underlying socket was closed and `WOLFSSL_BIO_FLAG_READ|WOLFSSL_BIO_FLAG_RETRY` was not set. The `wolfSSL_BIO_ctrl_pending(ssl->biord) == 0` is old and I can't find a reason to keep checking it. I left it just in the case where there is output data pending.

ZD 13606

julek-wolfssl committed:
Fix `BioReceive` for closed connection

The pending check was forcing a `WOLFSSL_CBIO_ERR_WANT_WRITE` return even though the underlying socket was closed and `WOLFSSL_BIO_FLAG_READ|WOLFSSL_BIO_FLAG_RETRY` was not set. The `wolfSSL_BIO_ctrl_pending(ssl->biord) == 0` is old and I can't find a reason to keep checking it. I left it just in the case where there is output data pending.

julek-wolfssl committed:
Ticket failure should result in a regular handshake

dgarske reviewed (approved):
Over to @SparkiDev for another review on the ticket change. Makes sense to me, but I'd like him to confirm.

SparkiDev commented on src/wolfio.c:
```diff
@@ -169,15 +169,17 @@ int BioReceive(WOLFSSL* ssl, char* buf, int sz, void* ctx)
     if (recvd <= 0) {
-        if (wolfSSL_BIO_supports_pending(ssl->biord) &&
+        if (/* ssl->biowr->wrIdx is checked for Bind9 */
+            ssl->biowr != NULL && ssl->biowr->type == WOLFSSL_BIO_BIO &&
```
Add wolfSSL_BIO_method_type() to do this line.

SparkiDev commented on src/wolfio.c:
```diff
@@ -169,15 +169,17 @@ int BioReceive(WOLFSSL* ssl, char* buf, int sz, void* ctx)
-        if (wolfSSL_BIO_supports_pending(ssl->biord) &&
+        if (/* ssl->biowr->wrIdx is checked for Bind9 */
+            ssl->biowr != NULL && ssl->biowr->type == WOLFSSL_BIO_BIO &&
+            ssl->biowr->wrIdx != 0 &&
```
Is this a check for pending write?
I believe the OpenSSL API is BIO_wpending(). So add wolfSSL_BIO_wpending().

SparkiDev commented on src/internal.c:
```diff
@@ -30757,7 +30757,9 @@ static int DoSessionTicket(WOLFSSL* ssl, const byte* input, word32* inOutIdx,
-        if (ret == WOLFSSL_TICKET_RET_FATAL || ret < 0) return ret;
+        if (ret == WOLFSSL_TICKET_RET_FATAL)
+            ret = WOLFSSL_TICKET_RET_REJECT;
+        if (ret < 0) return ret;
```
Move return statement to next line.
Thanks.

julek-wolfssl replied:
Moved.

julek-wolfssl committed:
Address code review

- Use functions instead of accessing `BIO` members
- Add `wolfSSL_BIO_method_type`

SparkiDev commented on src/ssl.c:
```diff
@@ -17590,6 +17590,11 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
         }
     }
+
+    int wolfSSL_BIO_method_type(const WOLFSSL_BIO *b)
```
I've moved all the wolfSSL_BIO functions to bio.c.
Move it there too. Thanks!

julek-wolfssl replied:
Moved to `bio.c`.
